### PR TITLE
Update ScriptErrorEvaluationFailed with DebugPlutusFailure

### PR DESCRIPTION
**Pull request by Jimbo4350**
Improved plutus script error debugging output. 

Before:
```
The Plutus script evaluation failed: An error has occurred:
The machine terminated because of an error, either from a built-in function or from an explicit use of 'error'.
Script debugging logs: PT5
```
After:
```
Command failed: transaction build  Error: The following scripts have execution failures:
the script for transaction input 0 (in ascending order of the TxIds) failed with: 
Script hash: 581cc61bfa1c138524b69f378bc69504322f39289ce554d549db4d1e2b50
Script language: PlutusV3
Protocol version: Version 10
Script arguments: 
   ScriptInfo: SpendingScript (TxOutRef {txOutRefId = cb2c9611ad51f6e37c529f7a3cb8dde4eaeaf97e5a47ba03a8c0026fe7a94bc5, txOutRefIdx = 0}) (Just (Datum {getDatum = I 0}))
   TxInfo:
     TxId: a95d7c79d606c74cde9aad2dae26db6ececa3915a6fc3075bea15ae8a8c69c86
     Inputs: [ cb2c9611ad51f6e37c529f7a3cb8dde4eaeaf97e5a47ba03a8c0026fe7a94bc5!0 -> - Value {getValue = Map {unMap = [(,Map {unMap = [("",5000000)]})]}} addressed to
                                                                                       ScriptCredential: c61bfa1c138524b69f378bc69504322f39289ce554d549db4d1e2b50 (no staking credential)
                                                                                       with datum
                                                                                       datum hash:  03170a2e7597b7b7e3d84c05391d139a62b157e78786d8c082f29dcf4c111314
                                                                                       with referenceScript
                                                                                        ]
     Reference inputs: []
     Outputs: [ - Value {getValue = Map {unMap = [(,Map {unMap = [("",2000000)]}),(c61bfa1c138524b69f378bc69504322f39289ce554d549db4d1e2b50,Map {unMap = [("MillarCoin",5)]})]}} addressed to
                  PubKeyCredential: e0d8d2f61fbf9c6b1a2db4b7073c50e6013e3374661a2403d0826920 (no staking credential)
                  with datum
                  no datum
                  with referenceScript

              , - Value {getValue = Map {unMap = [(,Map {unMap = [("",1000000)]})]}} addressed to
                  PubKeyCredential: e0d8d2f61fbf9c6b1a2db4b7073c50e6013e3374661a2403d0826920 (no staking credential)
                  with datum
                  datum hash:  ee155ace9c40292074cb6aff8c9ccdd273c81648ff1149ef36bcea6ebb8a3e25
                  with referenceScript

              , - Value {getValue = Map {unMap = [(,Map {unMap = [("",15000005000000)]})]}} addressed to
                  PubKeyCredential: 5b438ecb59a1fdd3b8ece9cf076821d827eb121460d027e7a9fdc351 (no staking credential)
                  with datum
                  no datum
                  with referenceScript
                   ]
     Fee: 0
     Value minted: Value {getValue = Map {unMap = [(c61bfa1c138524b69f378bc69504322f39289ce554d549db4d1e2b50,Map {unMap = [("MillarCoin",5)]})]}}
     TxCerts: [ TxCertRegStaking (ScriptCredential c61bfa1c138524b69f378bc69504322f39289ce554d549db4d1e2b50) (Just 400000) ]
     Wdrl: []
     Valid range: (-∞ , +∞)
     Signatories: []
     Redeemers: [ ( Spending (TxOutRef {txOutRefId = cb2c9611ad51f6e37c529f7a3cb8dde4eaeaf97e5a47ba03a8c0026fe7a94bc5, txOutRefIdx = 0})
                , 0 )
                , ( Minting c61bfa1c138524b69f378bc69504322f39289ce554d549db4d1e2b50
                , 0 )
                , ( Certifying 0 (TxCertRegStaking (ScriptCredential c61bfa1c138524b69f378bc69504322f39289ce554d549db4d1e2b50) (Just 400000))
                , 0 ) ]
     Datums: [ ( 03170a2e7597b7b7e3d84c05391d139a62b157e78786d8c082f29dcf4c111314
             , 0 )
             , ( ee155ace9c40292074cb6aff8c9ccdd273c81648ff1149ef36bcea6ebb8a3e25
             , 1 ) ]
     Votes: []
     Proposal Procedures: []
     Current Treasury Amount: 
     Treasury Donation: 
   Redeemer:
     0
Script evaluation error: An error has occurred:
The machine terminated because of an error, either from a built-in function or from an explicit use of 'error'.
Script execution logs:  PT5

Script base64 encoded arguments: 2Hmf2Hmfn9h5n9h5.....
Script base64 encoded bytes: WQ5wAQAAMjIzIjMiMyIyMjIyMjIyMjIyMjIy..... 
```

# Changelog

```yaml
- description: |
   Improve plutus script failure error
# uncomment types applicable to the change:
  type:
  - feature        # introduces a new feature
```

# Context

Additional context for the PR goes here. If the PR fixes a particular issue please provide a [link](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword=) to the issue.

# How to trust this PR

Highlight important bits of the PR that will make the review faster. If there are commands the reviewer can run to observe the new behavior, describe them.

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated. See [Running tests](https://github.com/input-output-hk/cardano-node-wiki/wiki/Running-tests) for more details
- [ ] Self-reviewed the diff

### CODE REVIEW

**Review by zliu41 — COMMENTED** [review]
Great, this is a nice improvement.

```
Script arguments: ScriptInfo: CertifyingScript 0 (TxCertRegStaking (ScriptCredential c61bfa1c138524b69f378bc69504322f39289ce554d549db4d1e2b50) (Just 400000))
TxInfo:
```

This makes it looks like "Script arguments" and "TxInfo" are at the same level. The following would be better:

```
Script arguments:
  argument 1:
    ScriptInfo: ...
    TxInfo: ...
  ...
```

Can you also print the base64-encoded arguments? That would be more useful for re-running the script locally. Or is it already printed elsewhere?

**Comment by Jimbo4350** [comment]
> Great, this is a nice improvement.
> 
> ```
> Script arguments: ScriptInfo: CertifyingScript 0 (TxCertRegStaking (ScriptCredential c61bfa1c138524b69f378bc69504322f39289ce554d549db4d1e2b50) (Just 400000))
> TxInfo:
> ```
> 
> This makes it look like "Script arguments" and "TxInfo" are at the same level. The following would be better:
> 
> ```
> Script arguments:
>   argument 1:
>     ScriptInfo: ...
>     TxInfo: ...
>   ...
> ```
> 
> Can you also print the base64-encoded arguments? That would be more useful for re-running the script locally. Or is it already printed elsewhere?

Made the changes however I'm not sure why it's saying `PT5` is an unknown error code.

**Comment by zliu41** [comment]
> Made the changes however I'm not sure why it's saying PT5 is an unknown error code.

I'm not sure. It doesn't look like this is from the plutus repo: grepping "unknown error code" returns no result.

**Comment by Jimbo4350** [comment]
> > Made the changes however I'm not sure why it's saying PT5 is an unknown error code.
> 
> I'm not sure. It doesn't look like this is from the plutus repo: grepping "unknown error code" returns no result.

It comes from: https://github.com/IntersectMBO/cardano-api/pull/672/files#diff-4909897d92b6ddafd851db4cfc8d8584ad03948c0d94a42a41c8a3391711cd48R73

**Comment by zliu41** [comment]
It's a bug in Plutus - I'll fix it.

By the way, I don't think you'd want to look up everything in that map and prints "unknown error code" if not found. The log may also contain arbitrary strings the user puts there.

**Comment by zliu41** [comment]
Fixed in https://github.com/IntersectMBO/plutus/pull/6635

File: cardano-api/internal/Cardano/Api/Plutus.hs [hunk 1]
```diff
@@ -0,0 +1,81 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Cardano.Api.Plutus
```
Comment (carbolymer) [inline cardano-api/internal/Cardano/Api/Plutus.hs:7]: Can you add haddock to this module, describing it's purpose?

File: cardano-api/internal/Cardano/Api/Plutus.hs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+import qualified PlutusLedgerApi.V1 as Plutus
+
+import qualified Data.ByteString.Base64 as B64
+import           Data.ByteString.Short as BSS
+import           Data.Text (Text)
+import qualified Data.Text as Text
+import qualified Data.Text.Encoding as Text
+import           Prettyprinter
+
+data DebugPlutusFailure
```
Comment (carbolymer) [inline cardano-api/internal/Cardano/Api/Plutus.hs:30]: Can you add haddocks, please?

File: cardano-api/internal/Cardano/Api/Plutus.hs [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+  = DebugPlutusFailure
+  { dpfEvaluationError :: Plutus.EvaluationError
+  , dpfScriptWithContext :: Plutus.PlutusWithContext L.StandardCrypto
+  , dpfExecutionUnits :: Plutus.ExUnits
+  , dpfExecutionLogs :: [Text]
+  }
+  deriving (Eq, Show)
+
+renderDebugPlutusFailure :: DebugPlutusFailure -> Text
```
Comment (carbolymer) [inline cardano-api/internal/Cardano/Api/Plutus.hs:39]: Maybe an instance of `Pretty` would be better suited here?

File: cardano-api/test/cardano-api-golden/Test/Golden/ErrorsSpec.hs [hunk 4]
```diff
@@ -270,10 +269,6 @@ test_ScriptExecutionError =
     [ ("ScriptErrorMissingTxIn", ScriptErrorMissingTxIn txin1)
     , ("ScriptErrorTxInWithoutDatum", ScriptErrorTxInWithoutDatum txin1)
     , ("ScriptErrorWrongDatum", ScriptErrorWrongDatum hashScriptData1)
-    ,
```
Comment (carbolymer) [inline cardano-api/test/cardano-api-golden/Test/Golden/ErrorsSpec.hs:273]: Why? Are we moving away from testing of rendering of error messages?

Also, if you're removing an entry from here, a golden file should get removed as well.

File: cardano-api/internal/Cardano/Api/Plutus.hs [hunk 5]
```diff
@@ -0,0 +1,81 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE UndecidableInstances #-}
```
Comment (carbolymer) [inline cardano-api/internal/Cardano/Api/Plutus.hs:5]: Are all of those extensions needed, especially `UndecidableInstances`?

**Review by carbolymer — APPROVED** [review]

File: cardano-api/internal/Cardano/Api/Plutus.hs (hunk 3 above)
Comment (Jimbo4350) [reply cardano-api/internal/Cardano/Api/Plutus.hs:46]: We actually have more render functions. We need to make a decision if we are going to rely on `Pretty` or these rendering functions (or both?).

File: cardano-api/test/cardano-api-golden/Test/Golden/ErrorsSpec.hs (hunk 4 above)
Comment (Jimbo4350) [reply cardano-api/test/cardano-api-golden/Test/Golden/ErrorsSpec.hs:273]: I was going to do this in a follow up PR but I'll do it now: https://github.com/IntersectMBO/cardano-api/pull/672/commits/b7a1d2a267931d8ec95ff0ae42cb98510b21c6e8

File: cardano-api/internal/Cardano/Api/Plutus.hs (hunk 5 above)
Comment (Jimbo4350) [reply cardano-api/internal/Cardano/Api/Plutus.hs:5]: Likely not. Result of a copy paste.

**Pull request merged**